pkg/store/bucket.go

```diff
@@ -1188,89 +1150,200 @@ func newBucketIndexReader(ctx context.Context, logger log.Logger, block *bucketB
 	return r
 }
 
-func (r *bucketIndexReader) preloadPostings() error {
-	const maxGapSize = 512 * 1024
+func (r *bucketIndexReader) lookupSymbol(o uint32) (string, error) {
+	s, ok := r.block.symbols[o]
+	if !ok {
+		return "", errors.Errorf("bucketIndexReader: unknown symbol offset %d", o)
+	}
+	return s, nil
+}
 
-	ps := r.loadedPostings
+// Postings returns postings in expanded list instead of index.Postings.
+// This is because we need to have them buffered anyway to perform efficient lookup
+// on object storage.
+// Found posting IDs (ps) are not strictly required to point to a valid Series, e.g. during
+// background garbage collections.
+func (r *bucketIndexReader) ExpandedPostings(ms []labels.Matcher) ([]uint64, error) {
+	var postingsToIntersect []index.Postings
 
-	sort.Slice(ps, func(i, j int) bool {
-		return ps[i].ptr.Start < ps[j].ptr.Start
-	})
-	parts := partitionRanges(len(ps), func(i int) (start, end uint64) {
-		return uint64(ps[i].ptr.Start), uint64(ps[i].ptr.End)
-	}, maxGapSize)
-	var g run.Group
+	// NOTE: Derived from tsdb.PostingsForMatchers.
+	for _, m := range ms {
+		matching, err := matchingLabels(r.LabelValues, m)
+		if err != nil {
+			return nil, errors.Wrap(err, "match postings")
+		}
+		if len(matching) == 0 {
+			continue
+		}
 
-	for _, p := range parts {
-		ctx, cancel := context.WithCancel(r.ctx)
-		i, j := p[0], p[1]
+		// We need to load all matching postings to tell what postings are intersecting with what.
+		postings, err := r.fetchPostings(matching)
+		if err != nil {
+			return nil, errors.Wrap(err, "get postings")
+		}
 
-		g.Add(func() error {
-			return r.loadPostings(ctx, ps[i:j], ps[i].ptr.Start, ps[j-1].ptr.End)
-		}, func(err error) {
-			if err != nil {
-				cancel()
-			}
-		})
+		postingsToIntersect = append(postingsToIntersect, postings)
 	}
 
-	if err := g.Run(); err != nil {
-		return err
+	if len(postingsToIntersect) == 0 {
+		return nil, nil
 	}
 
-	// TODO(bwplotka): Users are seeing lazyPostings not fully loaded. Double checking it here and printing more info
-	// on failure case.
-	for _, postings := range ps {
-		if postings.Postings != nil {
-			continue
-		}
+	ps, err := index.ExpandPostings(index.Intersect(postingsToIntersect...))
+	if err != nil {
+		return nil, errors.Wrap(err, "expand")
+	}
 
-		msg := fmt.Sprintf("found parts: %v bases on:", parts)
-		for _, p := range ps {
-			msg += fmt.Sprintf(" [start: %v; end: %v]", p.ptr.Start, p.ptr.End)
+	// As of version two all series entries are 16 byte padded. All references
+	// we get have to account for that to get the correct offset.
+	// We do it right at the beginning as it's easier than doing it more fine-grained
+	// at the loading level.
+	if r.block.indexVersion >= 2 {
+		for i, id := range ps {
+			ps[i] = id * 16
```

Comment:
not sure I understand this how does multiplying by 16 ensure its padded correctly? Maybe speak offline.

Reply:
That's 1:1 copied with what it was. 

Basically TSDB are now storing those reference every 16 bytes. The index format changed.

Reply:
Ahhhhh, okay now I understand we are calculating the offset which as we are saving everything as a fixed byte range we need to ensure we calculate the offset of id to point to the correct bytes :+1: